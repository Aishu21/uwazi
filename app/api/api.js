import bodyParser from 'body-parser'
import db_config from './config/database.js'

export default app => {

  //set db to use
  db_config.db_url = db_config.production;

  //common middlewares
  app.use(bodyParser.json());

  //module routes
  require('./auth/routes.js')(app);

}
