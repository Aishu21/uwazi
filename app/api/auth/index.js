import needsAuthorization from './authMiddleware';

export { needsAuthorization };
export { default as encryptPassowrd } from './encryptPassword';
export { comparePasswords } from './encryptPassword';
