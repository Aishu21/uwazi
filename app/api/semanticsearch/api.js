import fetch from 'isomorphic-fetch';
import SEMANTIC_SEARCH_URL from 'api/config/semanticsearch';

const semanticSearchAPI = {
  async processDocument(args) {
    const res = await fetch(`${SEMANTIC_SEARCH_URL}/semanticSearch/searchOneDoc`, {
      method: 'POST',
      body: JSON.stringify(args),
      headers: {
        'Content-Type': 'application/json'
      }
    });
    const body = await res.json();
    return body;
  }
};

export default semanticSearchAPI;
