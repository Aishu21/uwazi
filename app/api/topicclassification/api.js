/** @format */

import 'isomorphic-fetch';
import TOPIC_CLASSIFICATION_URL from 'api/config/topicclassification';

const topicClassificationAPI = {
  async processDocument({ seq, model }) {
    const url = new URL(TOPIC_CLASSIFICATION_URL);
    url.searchParams.set('model', model);
    const res = await fetch(TOPIC_CLASSIFICATION_URL, {
      method: 'POST',
      body: JSON.stringify(seq),
      headers: {
        'Content-Type': 'application/json',
      },
    });
    const body = await res.json();
    return body;
  },
};

export default topicClassificationAPI;
