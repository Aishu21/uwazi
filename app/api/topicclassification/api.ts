/** @format */

import 'isomorphic-fetch';
import isReachable from 'is-reachable';
import { tcServer } from 'api/config/topicClassification';
import { URL } from 'url';
import request from 'shared/JSONRequest';

export async function getModels() {
  const tcUrl = new URL('models', tcServer);
  const msTimeout = 1000;
  if (!(await isReachable(tcUrl.href, { timeout: msTimeout }))) {
    // TODO: move this backend check to server start-up time, maybe
    return {
      models: {},
      error: `Topic Classification server is unreachable (${msTimeout})`,
    };
  }
  return request
    .get(tcUrl.href)
    .then(async res => res.json)
    .catch(err => ({
      models: {},
      error: `Error from topic-classification server: ${err.toString()}`,
    }));
}

export async function checkModelReady(arg: { model: string }) {
  const tcUrl = new URL('models', tcServer);
  tcUrl.searchParams.set('model', arg.model);
  const msTimeout = 1000;
  if (!(await isReachable(tcUrl.href, { timeout: msTimeout }))) {
    // TODO: move this backend check to server start-up time, maybe
    return { models: {}, error: `Topic Classification server is unreachable (${msTimeout})` };
  }
  return request
    .get(tcUrl.href)
    .then(async res => res.json)
    .catch(err => ({
      models: '',
      error: `Error from topic-classification server: ${err.toString()}`,
    }));
}
