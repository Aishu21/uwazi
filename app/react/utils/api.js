import { browserHistory } from 'react-router';

import { isClient } from 'app/utils';
import { notify } from 'app/Notifications/actions/notificationsActions';
import { store } from 'app/store';
import loadingBar from 'app/App/LoadingProgressBar';

import { APIURL } from '../config.js';
import request from '../../shared/JSONRequest';

let cookie;
let locale;
let API_URL = APIURL;

const doneLoading = (data) => {
  loadingBar.done();
  return data;
};

const handleError = (e, endpoint) => {
  const error = e;
  error.endpoint = endpoint;

  if (!isClient) {
    return Promise.reject(error);
  }

  doneLoading();

  if (error.status === 401) {
    browserHistory.replace('/login');
  }

  if (error.status === 404) {
    browserHistory.replace('/404');
  }

  if (error.status === 500) {
    store.dispatch(notify('An error has occurred', 'danger'));
  }

  if (![500, 404, 401].includes(error.status)) {
    store.dispatch(notify(error.json.error, 'danger'));
  }

  return Promise.reject(error);
};

const _request = (url, data, method) => {
  loadingBar.start();
  return request[method](API_URL + url, data, { 'Content-language': locale, Cookie: cookie, 'X-Requested-With': 'XMLHttpRequest' })
  .then(doneLoading)
  .catch(e => handleError(e, { url, method }));
};

export default {
  get: (url, data) => _request(url, data, 'get'),

  post: (url, data) => _request(url, data, 'post'),

  delete: (url, data) => _request(url, data, 'delete'),

  cookie(c) {
    cookie = c;
  },

  locale(key) {
    locale = key;
  },

  APIURL(url) {
    API_URL = url;
  }
};
