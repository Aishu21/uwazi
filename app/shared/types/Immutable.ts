/** @format */

export type IImmutable<T> = T extends string
  ? string
  : T extends number
  ? number
  : T extends boolean
  ? boolean
  : T extends null
  ? null
  : T extends undefined
  ? undefined
  : T extends Array<infer Elem>
  ? {
      toJS(): Elem[];
      get(i: number): IImmutable<Elem>;
    }
  : {
      toJS(): T;
      get<Field extends keyof T>(_field: Field): IImmutable<T[Field]>;
    };
