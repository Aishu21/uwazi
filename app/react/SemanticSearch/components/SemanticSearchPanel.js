import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Field, LocalForm, actions } from 'react-redux-form';

import { t } from 'app/I18N';
import { Icon, } from 'UI';
import ShowIf from 'app/App/ShowIf';

import api from '../SemanticSearchAPI';

import SidePanel from 'app/Layout/SidePanel';
import SearchList from './SearchList';

export class SemanticSearchSidePanel extends Component {
  constructor(props) {
    super(props);
    this.state = {
      page: 'main',
      searches: null
    };

    this.onSubmit = this.onSubmit.bind(this);
    this.submitForm = this.submitForm.bind(this);
    this.showMainPage = this.showMainPage.bind(this);
    this.showNewSearchPage = this.showNewSearchPage.bind(this);
  }

  componentDidMount() {
    if (!this.searches) {
      api.getAllSearches().then(res => this.setState({ searches: res }));
    }
  }

  onSubmit(model) {
    const { currentSearch, selectedDocuments } = this.props;
    const { searchTerm } = model;
    const documents = selectedDocuments ?
      selectedDocuments.toJS().map(doc => doc.sharedId) : [];
    api.search({
      searchTerm,
      documents,
      query: currentSearch
    });
  }

  attachDispatch(dispatch) {
    this.formDispatch = dispatch;
  }

  showMainPage() {
    this.setState({ page: 'main' });
  }

  showNewSearchPage() {
    this.setState({ page: 'new' });
  }

  submitForm() {
    this.formDispatch(actions.submit('searchText'));
  }

  render() {
    const { page, searches } = this.state;
    return (
      <SidePanel open={this.props.open} className="metadata-sidepanel semantic-search">
        <ShowIf if={page === 'new'}>
          <div className="sidepanel-footer">
            <span
              className="btn btn-danger cancel-search"
              onClick={this.showMainPage}
            >
              <Icon icon="times" />
              <span className="btn-label">{t('System', 'Cancel')}</span>
            </span>
            <button
              className="btn btn-success"
              onClick={this.submitForm}
            >
              <Icon icon="search" />
              <span className="btn-label">{t('System', 'Start search')}</span>
            </button>
          </div>
        </ShowIf>
        <div className="sidepanel-body">
          <p className="sidepanel-title">{t('System', 'Semantic search')}</p>
          <ShowIf if={page === 'main'}>
            <div style={{
              display: 'flex',
              alignItems: 'center',
              justifyContent: 'center',
              marginTop: 30,
              marginBottom: 30 }}
            >
              <button
                className="btn btn-default new-search"
                onClick={this.showNewSearchPage}
              >
                {t('System', 'Start new search')}
              </button>
            </div>
          </ShowIf>
          <ShowIf if={page === 'main' && !!searches}>
            <SearchList searches={searches} />
          </ShowIf>
          <ShowIf if={page === 'new'}>
            <div>
              <LocalForm
                model="searchText"
                autoComplete="off"
                onSubmit={this.onSubmit}
                getDispatch={dispatch => this.attachDispatch(dispatch)}
              >
                <div className="search-box">
                  <div className="input-group">
                    <Field model=".searchTerm">
                      <Icon icon="search" />
                      <input
                        type="text"
                        placeholder={t('System', 'Search', null, false)}
                        className="form-control"
                        autoComplete="off"
                      />
                      <Icon icon="times"/>
                    </Field>
                  </div>
                </div>
              </LocalForm>
            </div>
          </ShowIf>
        </div>
      </SidePanel>
    );
  }
}

SemanticSearchSidePanel.propTypes = {
  currentSearch: PropTypes.object.isRequired,
  selectedDocuments: PropTypes.object.isRequired,
  open: PropTypes.bool.isRequired
};

export function mapStateToProps(state, props) {
  return {
    currentSearch: state[props.storeKey].search,
    selectedDocuments: state[props.storeKey].ui.get('selectedDocuments'),
    open: state[props.storeKey].ui.get('showSemanticSearchPanel')
  };
}

export default connect(mapStateToProps)(SemanticSearchSidePanel);
