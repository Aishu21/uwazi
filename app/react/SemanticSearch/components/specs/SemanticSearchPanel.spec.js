import React from 'react';
import { shallow } from 'enzyme';
import Immutable from 'immutable';

import { SemanticSearchSidePanel, mapStateToProps } from '../SemanticSearchPanel';
import api from '../../SemanticSearchAPI';

describe('SemanticSearchPanel', () => {
  let initialState;
  let initialProps;
  beforeEach(() => {
    initialState = {
      library: {
        search: { query: 'library search'},
        ui: {
          selectedDocuments: [],
          showSemanticSearchPanel: true
        }
      }
    };
    initialProps = {
      storeKey: 'library'
    };
    jest.spyOn(api, 'getAllSearches').mockResolvedValue([
      { searchTerm: 'search1' },
      { searchTerm: 'search2' }
    ]);
  });

  const getProps = () => {
    const library = {
      ...initialState.library,
      ui: Immutable.fromJS(initialState.library.ui)
    };
    const state = {
      ...initialState,
      library
    };
    return mapStateToProps(state, initialProps);
  };
  const render = () => shallow(<SemanticSearchSidePanel {...getProps()} />);
  it('should fetch all searches when mounted', (done) => {
    render();
    setImmediate(() => {
      expect(api.getAllSearches).toHaveBeenCalled();
      done();
    });
  });
  it('should render SearchList with fetched searches', (done) => {
    const component = render();
    setImmediate(() => {
      expect(component.update()).toMatchSnapshot(done);
      done();
    });
  });
  it('should render new search form when search button is clicked', () => {
    const component = render();
    component.find('.new-search').simulate('click');
    expect(component.update()).toMatchSnapshot();
  });
  describe('when on new search form', () => {
    it('should go back to search list page when the cancel button is clicked', () => {
      const component = render();
      component.setState({ page: 'new', searches: [] });
      component.find('.cancel-search').simulate('click');
      expect(component.update()).toMatchSnapshot();
    });
  });
});
