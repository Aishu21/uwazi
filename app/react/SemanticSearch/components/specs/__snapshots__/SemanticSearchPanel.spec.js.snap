// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SemanticSearchPanel should render SearchList with fetched searches 1`] = `
<SidePanel
  className="metadata-sidepanel semantic-search"
  open={true}
>
  <ShowIf
    if={false}
  >
    <div
      className="sidepanel-footer"
    >
      <span
        className="btn btn-danger cancel-search"
        onClick={[Function]}
      >
        <Component
          icon="times"
        />
        <span
          className="btn-label"
        >
          Cancel
        </span>
      </span>
      <button
        className="btn btn-success"
        onClick={[Function]}
      >
        <Component
          icon="search"
        />
        <span
          className="btn-label"
        >
          Start search
        </span>
      </button>
    </div>
  </ShowIf>
  <div
    className="sidepanel-body"
  >
    <p
      className="sidepanel-title"
    >
      Semantic search
    </p>
    <ShowIf
      if={true}
    >
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "justifyContent": "center",
            "marginBottom": 30,
            "marginTop": 30,
          }
        }
      >
        <button
          className="btn btn-default new-search"
          onClick={[Function]}
        >
          Start new search
        </button>
      </div>
    </ShowIf>
    <ShowIf
      if={true}
    >
      <Connect(SearchList)
        searches={
          Array [
            Object {
              "searchTerm": "search1",
            },
            Object {
              "searchTerm": "search2",
            },
          ]
        }
      />
    </ShowIf>
    <ShowIf
      if={false}
    >
      <div>
        <LocalForm
          autoComplete="off"
          getDispatch={[Function]}
          initialState={Object {}}
          model="searchText"
          onSubmit={[Function]}
        >
          <div
            className="search-box"
          >
            <div
              className="input-group"
            >
              <Modeled(Connect(Field))
                model=".searchTerm"
              >
                <Component
                  icon="search"
                />
                <input
                  autoComplete="off"
                  className="form-control"
                  placeholder="Search"
                  type="text"
                />
                <Component
                  icon="times"
                />
              </Modeled(Connect(Field))>
            </div>
          </div>
        </LocalForm>
      </div>
    </ShowIf>
  </div>
</SidePanel>
`;

exports[`SemanticSearchPanel should render new search form when search button is clicked 1`] = `
<SidePanel
  className="metadata-sidepanel semantic-search"
  open={true}
>
  <ShowIf
    if={true}
  >
    <div
      className="sidepanel-footer"
    >
      <span
        className="btn btn-danger cancel-search"
        onClick={[Function]}
      >
        <Component
          icon="times"
        />
        <span
          className="btn-label"
        >
          Cancel
        </span>
      </span>
      <button
        className="btn btn-success"
        onClick={[Function]}
      >
        <Component
          icon="search"
        />
        <span
          className="btn-label"
        >
          Start search
        </span>
      </button>
    </div>
  </ShowIf>
  <div
    className="sidepanel-body"
  >
    <p
      className="sidepanel-title"
    >
      Semantic search
    </p>
    <ShowIf
      if={false}
    >
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "justifyContent": "center",
            "marginBottom": 30,
            "marginTop": 30,
          }
        }
      >
        <button
          className="btn btn-default new-search"
          onClick={[Function]}
        >
          Start new search
        </button>
      </div>
    </ShowIf>
    <ShowIf
      if={false}
    >
      <Connect(SearchList)
        searches={null}
      />
    </ShowIf>
    <ShowIf
      if={true}
    >
      <div>
        <LocalForm
          autoComplete="off"
          getDispatch={[Function]}
          initialState={Object {}}
          model="searchText"
          onSubmit={[Function]}
        >
          <div
            className="search-box"
          >
            <div
              className="input-group"
            >
              <Modeled(Connect(Field))
                model=".searchTerm"
              >
                <Component
                  icon="search"
                />
                <input
                  autoComplete="off"
                  className="form-control"
                  placeholder="Search"
                  type="text"
                />
                <Component
                  icon="times"
                />
              </Modeled(Connect(Field))>
            </div>
          </div>
        </LocalForm>
      </div>
    </ShowIf>
  </div>
</SidePanel>
`;

exports[`SemanticSearchPanel when on new search form should go back to search list page when the cancel button is clicked 1`] = `
<SidePanel
  className="metadata-sidepanel semantic-search"
  open={true}
>
  <ShowIf
    if={false}
  >
    <div
      className="sidepanel-footer"
    >
      <span
        className="btn btn-danger cancel-search"
        onClick={[Function]}
      >
        <Component
          icon="times"
        />
        <span
          className="btn-label"
        >
          Cancel
        </span>
      </span>
      <button
        className="btn btn-success"
        onClick={[Function]}
      >
        <Component
          icon="search"
        />
        <span
          className="btn-label"
        >
          Start search
        </span>
      </button>
    </div>
  </ShowIf>
  <div
    className="sidepanel-body"
  >
    <p
      className="sidepanel-title"
    >
      Semantic search
    </p>
    <ShowIf
      if={true}
    >
      <div
        style={
          Object {
            "alignItems": "center",
            "display": "flex",
            "justifyContent": "center",
            "marginBottom": 30,
            "marginTop": 30,
          }
        }
      >
        <button
          className="btn btn-default new-search"
          onClick={[Function]}
        >
          Start new search
        </button>
      </div>
    </ShowIf>
    <ShowIf
      if={true}
    >
      <Connect(SearchList)
        searches={Array []}
      />
    </ShowIf>
    <ShowIf
      if={false}
    >
      <div>
        <LocalForm
          autoComplete="off"
          getDispatch={[Function]}
          initialState={Object {}}
          model="searchText"
          onSubmit={[Function]}
        >
          <div
            className="search-box"
          >
            <div
              className="input-group"
            >
              <Modeled(Connect(Field))
                model=".searchTerm"
              >
                <Component
                  icon="search"
                />
                <input
                  autoComplete="off"
                  className="form-control"
                  placeholder="Search"
                  type="text"
                />
                <Component
                  icon="times"
                />
              </Modeled(Connect(Field))>
            </div>
          </div>
        </LocalForm>
      </div>
    </ShowIf>
  </div>
</SidePanel>
`;
